In the previous notebook, we learned how to do Simple Linear Regression. It's called simple because there is only one independent variable and one dependent variable. There's absolutely nothing in principle that prevents us from using multiple independent variables. In this case, it's called Multiple Linear Regression, but it uses the same scikit-learn function. The function that Multiple Linear Regression is attempting to estimate looks like this:

$$ y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \beta_3 x_3 + ... + \beta_n x_n$$

We simply have a new coefficient for each new feature in the independent variables. 

Let's create some synthetic data. There will be 4 independent variables. The dependent variable will be created as follows:

$$ y = 10 + 0.5x_1 + (-0.7)x_2 + 12x_3 + (-4.2)x_4 + \epsilon$$

As usually, the $\epsilon$ represents noise that we're adding to the data.

First, let's import what we need.

In [1]:
from sklearn.linear_model import LinearRegression
import numpy as np

np.random.seed(9216) # setting a seed for consistent results

# Generating Synthetic Data

Now, we'll make 1000 samples. The $x$'s themselves will be distributed normally according to some mean and standard deviation that I'm just making up. 

In [2]:
x1 = np.random.normal(5, 10, 1000)
x2 = np.random.normal(-2.45, 4, 1000)
x3 = np.random.normal(57.8, 14, 1000)
x4 = np.random.normal(31.4, 20, 1000)
noise = np.random.normal(0, 10, 1000)

y = 10 + 0.5*x1 - 0.7*x2 + 12*x3 - 4.2*x4 + noise

The ``x`` variables are our independent variables. The model will attempt to learn from this data. ``y`` is our dependent variable. We know *how* it was made, but the model does not. 

Scikit learn ``fit`` functions expects that the dependent variable training data by packaged as an array of arrays. But right now, each of our ``x`` variables are in separate arrays. We need to merge them together somehow. 

In [3]:
X = np.array([x1, x2, x3, x4])

This new array, ``X``, has 4 arrays inside of it. Each array has a length of 1000. We can therefor consider this to be a matrix of size $4 \times 1000$. Each row is one of the data features, each column is one of the samples. We can actually check the size of this array of arrays by using dot notation and seeing the value of its attribute ``shape``.


In [4]:
X.shape

(4, 1000)

Indeed, ``X`` is a $4 \times 1000$ grid. 

Recall, however, that in both programming and the math of data mining the training data should be matrices of size $n \times d$, that is, the number of samples by the number of features. Right now, ``X`` is backwards. We need to flip the rows and the columns. I'm going to do that using a numpy function called ``transpose``.

In [5]:
X = np.transpose(X)
X.shape

(1000, 4)

``X`` is an array of arrays, but ``y`` is still an array of numbers. We'll use the ``reshape`` trick to turn it into an array of arrays.

In [6]:
y[:5]

array([ 484.39763051,  347.86637141,  653.44863166,  275.12867872,
        515.38212337])

In [7]:
y.shape

(1000,)

In [8]:
y = y.reshape(-1, 1)
y[:5]

array([[ 484.39763051],
       [ 347.86637141],
       [ 653.44863166],
       [ 275.12867872],
       [ 515.38212337]])

In [9]:
y.shape

(1000, 1)

# Making and Training the Model

Now we're good to go. Let's make the regressor model. First, we call the factory method ``LinearRegression`` to initialize an empty, new regressor model.

In [10]:
lr = LinearRegression()

Then, we train the model by calling ``fit`` and providing it with the training data and the outputs.

In [11]:
lr.fit(X, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [16]:
print(y[:5])

[[ 484.39763051]
 [ 347.86637141]
 [ 653.44863166]
 [ 275.12867872]
 [ 515.38212337]]


And that's it! The coefficients are in the model, ``lr``. Let's take a look first at the intercept.

In [13]:
lr.intercept_

array([ 11.82809212])

And the rest of the coefficients.

In [14]:
lr.coef_

array([[  0.44653135,  -0.67876566,  11.98323056,  -4.21864062]])

Remember, the function we used to synthesize the data was 

$$ y = 10 + 0.5x_1 + (-0.7)x_2 + 12x_3 + (-4.2)x_4 + \epsilon$$

We've looked at the trained models coefficients, and so we know the estimation it came up with is

$$ y = 11.83 + 0.45x_1 + (-0.68)x_2 + 11.98x_3 + (-4.22) x_4$$

Not bad. As with Simple Regression, now that we have a model we can predict new ``y`` values for new, unseen data. Let's say I have observed 2 new instances. Each instance has 4 features, $x_1$ through $x_4$. I simply call the predict function, and the model tells me its estimates of $y$.

In [15]:
lr.predict([[1, 2, 3, 4], [-5, 4, 7, 35]])

array([[ 29.99222134],
       [-56.8894352 ]])